zookeeper/source/zk_nn_client.cc

```diff
@@ -0,0 +1,952 @@
+			FileZNode znode_data;
+			read_file_znode(znode_data, path);
+
+			// set the file status in the get file info response res
+			HdfsFileStatusProto* status = res.mutable_fs();
+
+			set_file_info(status, path, znode_data);
+			LOG(INFO) << CLASS_NAME << "Got info for file ";
+			return;
+		}
+		LOG(INFO) << CLASS_NAME << "No file to get info for";
+	}
+
+	/**
+	 * Create a node in zookeeper corresponding to a file
+	 */
+	int ZkNnClient::create_file_znode(const std::string& path, FileZNode* znode_data) {
+		int error_code;
+		if (!file_exists(path)) {
+			LOG(INFO)<< "Creating file znode at " << path;
+			{
+				LOG(INFO) << CLASS_NAME << znode_data->replication;
+				LOG(INFO) << CLASS_NAME << znode_data->owner;
+				LOG(INFO) << CLASS_NAME << "size of znode is " << sizeof(*znode_data);
+			}
+			// serialize struct to byte vector
+			std::vector<std::uint8_t> data(sizeof(*znode_data));
+			file_znode_struct_to_vec(znode_data, data);
+			// crate the node in zookeeper
+			if (!zk->create(ZookeeperPath(path), data, error_code)) {
+				LOG(ERROR) << CLASS_NAME << "Create failed" << error_code;
+				return 0;
+				// TODO : handle error
+			}
+			return 1;
+		}
+		return 0;
+	}
+
+	void ZkNnClient::delete_node_wrapper(std::string& path, DeleteResponseProto& response) {
+		int error_code;
+		if (!zk->delete_node(ZookeeperPath(path), error_code)) {
+			response.set_result(false);
+			LOG(ERROR) << CLASS_NAME << "Error deleting node at " << path << " because of error = " << error_code;
+			return;
+		}
+		LOG(INFO) << CLASS_NAME << "Successfully deletes znode";
+	}
+
+
+	/**
+	 * Go down directories recursively. If a child is a file, then put its deletion on a queue.
+	 * Files delete themselves, but directories are deleted by their parent (so root can't be deleted)
+	 */
+	void ZkNnClient::destroy(DeleteRequestProto& request, DeleteResponseProto& response) {
+
+		int error_code;
+		const std::string& path = request.src();
+		bool recursive = request.recursive();
+		response.set_result(true);
+		if (!file_exists(path)) {
+			return response.set_result(false);
+		}
+		FileZNode znode_data;
+		read_file_znode(znode_data, path);
+
+		// we have a directory
+		if (recursive) {
+			std::vector<std::string> children;
+			if (!zk->get_children(ZookeeperPath(path), children, error_code)) {
+				LOG(INFO) << CLASS_NAME << "Could not get children for " << path << " because of error = " << error_code;
+				response.set_result(false);
+				// zlock.unlock();
+				return;
+			}
+			// delete the kids
+			for (auto src : children) {
+				DeleteRequestProto request_child;
+				DeleteResponseProto response_child;
+				request_child.set_src(path + "/" + src);
+				request_child.set_recursive(true);
+				destroy(request_child, response_child);
+				if (response_child.result() == false) { // propogate failures updwards
+					response.set_result(false);
+				}
+			}
+		}
+		else if (znode_data.filetype == IS_DIR || znode_data.filetype == IS_DIR1) {
```

Comment:
@Vegetable26 DIR1 is actually a lingering error from not writing to the fileSystem root node and settings its directory information (so the info would be zerod out, although not definitively). We write to the root node now, so this shouldn't be necessary.